drivers/timer/arcv2_timer0.c

```diff
@@ -279,29 +345,51 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
 
 #if defined(CONFIG_TICKLESS_KERNEL)
 	u32_t delay;
+	u32_t unannounced;
 
-	ticks = MIN(MAX_TICKS, MAX(ticks - 1, 0));
-
-	/* Desired delay in the future */
-	delay = (ticks == 0) ? MIN_DELAY : ticks * CYC_PER_TICK;
+	ticks = MIN(MAX_TICKS, (u32_t)(MAX((s32_t)(ticks - 1), 0)));
 
 	k_spinlock_key_t key = k_spin_lock(&lock);
 
-	delay += elapsed();
 
-	/* Round delay up to next tick boundary */
-	delay = ((delay + CYC_PER_TICK - 1) / CYC_PER_TICK) * CYC_PER_TICK;
+	cycle_count += elapsed();
+	overflow_cyc = 0U;
+
 
-	if (last_load != delay) {
-		if (timer0_control_register_get() & _ARC_V2_TMR_CTRL_IP) {
-			delay -= last_load;
+	/* normal case */
+	unannounced = cycle_count - announced_cycles;
+
+	if ((s32_t)unannounced < 0) {
+		/* We haven't announced for more than half the 32-bit
+		 * wrap duration, because new timeouts keep being set
+		 * before the existing one fires. Force an announce
+		 * to avoid loss of a wrap event, making sure the
+		 * delay is at least the minimum delay possible.
+		 */
+		last_load = MIN_DELAY;
+	} else {
+		/* Desired delay in the future */
+		delay = ticks * CYC_PER_TICK;
+
+		/* Round delay up to next tick boundary */
+		delay += unannounced;
+		delay =
+		 ((delay + CYC_PER_TICK - 1) / CYC_PER_TICK) * CYC_PER_TICK;
+
+		delay -= unannounced;
+		delay = MAX(delay, MIN_DELAY);
+
+		if (delay > MAX_CYCLES) {
+			last_load = MAX_CYCLES;
+		} else {
+			last_load = delay;
 		}
-		timer0_limit_register_set(delay - 1);
-		last_load = delay;
-		timer0_control_register_set(_ARC_V2_TMR_CTRL_NH |
-							 _ARC_V2_TMR_CTRL_IE);
 	}
 
+	timer0_limit_register_set(last_load - 1);
+	timer0_count_register_set(0);
```

Comment:
Could reprogramming the count register be avoided in some way? My concern is timer drift. There is some (variable) time between the read of count in elapsed() and the setting to 0 here. The earlier solution avoided reprogramming of the count register. Alternatively, can the call to timer0_count_register_set(0) be moved up, just after elapsed()? That reduces the amount of cycles between reading and resetting.

Reply:
I have ever considered about this, based on current mechanism, difficult to do that.